firedrake/projection.py

```diff
@@ -97,3 +97,41 @@ def project(v, V, bcs=None, mesh=None,
            solver_parameters=solver_parameters,
            form_compiler_parameters=form_compiler_parameters)
     return ret
+
+
+class Projector(object):
+    """
+    A projector projects a UFL expression into a function space
+    and places the result in a function from that function space,
+    allowing the solver to be reused.
+
+    :arg v: the :class:`.Expression`, :class:`ufl.Expr` or
+         :class:`.Function` to project
```

Comment:
This implementation doesn't support projecting Expression objects.  You can't use them directly in a form.